backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/ModularContractService.java

```diff
@@ -0,0 +1,124 @@
+import de.metas.contracts.modular.settings.ModularContractSettingsDAO;
+import de.metas.contracts.modular.settings.ModularContractType;
+import de.metas.contracts.modular.settings.ModuleConfig;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+@Service
+public class ModularContractService
+{
+	public enum ModelAction
+	{
+		COMPLETED, REVERSED, REACTIVATED
+	}
+
+	private final ModularContractLogDAO contractLogDAO;
+	private final ModularContractHandlerFactory modularContractHandlerFactory;
+	private final ModularContractSettingsDAO modularContractSettingsDAO;
+	private final IFlatrateDAO flatrateDAO = Services.get(IFlatrateDAO.class);
+
+	public ModularContractService(
+			@NonNull final ModularContractLogDAO contractLogDAO,
+			@NonNull final ModularContractHandlerFactory modularContractHandlerFactory,
+			@NonNull final ModularContractSettingsDAO modularContractSettingsDAO)
+	{
+		this.contractLogDAO = contractLogDAO;
+		this.modularContractHandlerFactory = modularContractHandlerFactory;
+		this.modularContractSettingsDAO = modularContractSettingsDAO;
+	}
+
+	public void invokeWithModel(@NonNull final Object model, @NonNull final ModelAction action)
+	{
+		modularContractHandlerFactory.getApplicableHandlersFor(model)
+				.forEach(handler -> invokeWithModel(handler, model, action));
+	}
+
+	private <T> void invokeWithModel(@NonNull final IModularContractTypeHandler<T> handler, final @NonNull T model, final @NonNull ModelAction action)
+	{
+		handler.getContractIds(model)
+				.filter(flatrateTermId -> isApplicableContract(handler, flatrateTermId))
+				.forEach(flatrateTermId -> invokeWithModel(handler, model, action, flatrateTermId));
+	}
+
+	private <T> boolean isApplicableContract(@NonNull final IModularContractTypeHandler<T> handler, @NonNull final FlatrateTermId flatrateTermId)
+	{
+		if (!isModularContract(flatrateTermId))
+		{
+			return false;
+		}
+
+		final ModularContractSettings settings = modularContractSettingsDAO.getByFlatrateTermIdOrNull(flatrateTermId);
+		return isHandlerApplicableForSettings(handler, settings);
+	}
+
+	private boolean isModularContract(@NonNull final FlatrateTermId flatrateTermId)
+	{
+		final I_C_Flatrate_Term flatrateTerm = flatrateDAO.retrieveTerm(flatrateTermId);
+		return Objects.equals(X_C_Flatrate_Term.TYPE_CONDITIONS_ModularContract, flatrateTerm.getType_Conditions());
+	}
+
+	private <T> boolean isHandlerApplicableForSettings(@NonNull final IModularContractTypeHandler<T> handler, @Nullable final ModularContractSettings settings)
+	{
+		if (settings == null)
+		{
+			return false;
+		}
+		final String handlerClassName = handler.getClass().getName();
+		return settings.getModuleConfigs()
+				.stream()
+				.map(ModuleConfig::getModularContractType)
+				.map(ModularContractType::getClassName)
+				.anyMatch(handlerClassName::equals);
```

Comment:
i find this logic a bit twisted / abusing steams API

IMHO i would rewrite it as:
```
settings.getModuleConfigs()
.stream()
.anyMatch(config -> config.isMatchingClassName(handlerClassName));
```
and I would introduce ModuleCOnfig.isMatchingClassName(String className) method.

Reply:
done